Support for MPI communicators which are provided externally

aa-g opened the pull request:
## Proposed Changes
Replace references to the global MPI communicator `MPI_COMM_WORLD` with `SU2_MPI::GetComm()` where appropriate.

This prevents deadlocking in cases where SU2 is given an MPI communicator other than `MPI_COMM_WORLD`, for example when creating an instance of `CDriver` through the Python interface.


## Related Work
Issue #1179 


## PR Checklist
- [X] I am submitting my contribution to the develop branch.
- [X] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags, or simply --warnlevel=2 when using meson).
- [X] My contribution is commented and consistent with SU2 style.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

pcarruscag commented on Common/include/linear_algebra/CPastixWrapper.hpp:
```diff
@@ -93,7 +93,7 @@ class CPastixWrapper
    */
   void Run() {
-    dpastix(&state, MPI_COMM_WORLD, nCols, colptr.data(), rowidx.data(), values.data(),
+    dpastix(&state, SU2_MPI::GetComm(), nCols, colptr.data(), rowidx.data(), values.data(),
```
I know it is a bit nit picky but can we rename GetComm() to just Comm(), just to make things shorter.

aa-g replied:
I agree that the proposed changes make the code slightly less readable.

Unfortunately, the name `SU2_MPI::Comm` is already defined in mpi_structure.hpp, line 98, as a `typedef` to `MPI_Comm`. Maybe this would have to be renamed ?

This change would also break consistency with other methods of `SU2_MPI`, such as `GetRank()` and `GetSize()`.

pcarruscag replied:
Hmmm alright consistency is important, especially since we don't have a lot of it.

pcarruscag reviewed (commented):
Thank you for opening the PR!

Delete line 470 of mpi_structure.hpp please.

aa-g commented:
> Thank you for opening the PR!
> 
> Delete line 470 of mpi_structure.hpp please.

No worries. It took a while to get it going due to some issues on my side catching up with 7.1.0 (in particular, compiling against MPICH).

Regarding the deletion of the dummy `MPI_COMM_WORLD` from mpi\_structure.hpp: the value is still referenced in several places, namely line 32 of mpi\_structure.cpp, and in each of the SU2\_CFD.cpp, SU2\_DEF.cpp, SU2\_DOT.cpp, SU2\_GEO.cpp, SU2\_SOL.cpp, where `MPI_COMM_WORLD` would be the default communicator.

pcarruscag commented:
Add a method to the MPI wrappers to get the global communicator or something equivalent, the only way to make this kind of change future-compatible is to break compilation.

pcarruscag commented:
I merged #1178 earlier and fixed the old and new conflicts, some MPI_COMM_WORLD's will have been re-introduced in CFVMFlowSolverBase.hpp and .inl because in my previous 2 PR's I moved a lot of code.

aa-g commented:
I've removed the dummy `MPI_COMM_WORLD` altogether now. To keep the changes minimal, I've just changed the way that `CBaseMPIWrapper::currentComm` is initialised in `mpi_structure.cpp`, rather than introducing a new method (which would only be used once, at this location).

Now, the `CBaseMPIWrapper::currentComm` is initialised to `MPI_COMM_WORLD` if compiling with MPI, otherwise it is set to the dummy value `0`, consistent with the previous behaviour.

pcarruscag commented on Common/include/parallelization/mpi_structure.cpp:
```diff
@@ -27,9 +27,17 @@
 int CBaseMPIWrapper::Rank = 0;
 int CBaseMPIWrapper::Size = 1;
+
+/* Set the default MPI Communicator */
+#ifdef HAVE_MPI
 CBaseMPIWrapper::Comm CBaseMPIWrapper::currentComm = MPI_COMM_WORLD;
+#else
+CBaseMPIWrapper::Comm CBaseMPIWrapper::currentComm = 0;  // dummy value
+#endif
```
I like this solution.
The rank and size are updated by the SetComm function so everything should work ok when creating a driver from scratch.

aa-g replied:
Glad to hear that. Is anything still missing ? Not sure how to go about the failing check for code complexity ...

pcarruscag replied:
Not for me, I'll ask around the developers meeting tomorrow if anyone else has comments, codefactor glitches out sometimes no need to do anything.

By the way would you like to be added to su2code? You can then have branches here instead of in your fork, makes it simpler to manage future PR's you might open.

aa-g replied:
Yes, I suppose it would be more convenient that way. Thanks !
